lib/transaction.c

```diff
@@ -633,7 +633,17 @@ assert(otherFi != NULL);
 		rConflicts = handleColorConflict(ts, fs, fi, i,
 						otherFs, otherFi, otherFileNum);
 
-		if (rConflicts && reportConflicts) {
+		/*
+		 * This may be a false positive (two separate files) if a
+		 * symlink is being replaced by a directory in one of these two
+		 * paths.  This check extends the one for removal conflicts in
+		 * handleInstInstalledFile().
+		 */
+		int reportThis = !(p == otherTe &&
+				   rpmtsFlags(ts) & RPMTRANS_FLAG_TEST &&
+				   rpmteHaveTransScript(p, RPMTAG_PRETRANS));
```

Comment:
Mostly stylistic, but I'd prefer doing this similar to the related check in handleInstInstalledFile(), set rConflicts to 0 on the heuristics rather than add a new variable to the scene. You could even add a tiny helper function for the common part of the heuristics, which would help keep them in sync across any future changes.

Reply:
Good point. Also, the common function is something I actually wanted to do, but couldn't come up with a descriptive-enough (yet not super long) name for it :smile:
But I'll give it another try, as I agree it would be the best way to functionally and semantically "connect" those two checks.

Reply:
If descriptive names fail, try the opposite: short and incomprehensible, just commented to explain the purpose :sweat_smile:

Reply:
Point being, don't get hung up on a name. If lacking better ideas, isTransientConflict() would work for me, just move the lengthy comment on the %pretrans hack from handleInstInstalledFile() to comment the new function instead.

Reply:
Hmm. Wouldn't this now apply the heuristic to *all* file conflicts on newly installed files? The "removal conflict" case specifically looks at the dir <-> link cases and the %pretrans heuristic should only ever be applied to those here too.

Reply:
Unfortunately, that's the case, yes. I contemplated adding a specific check for the link -> dir case, but I thought it would be a bit more complicated, since you would have to iterate through each path component and check what its original vs. new status is. It could certainly be done and might not even be that difficult after all, but there's one catch that immediately comes to mind, which is - what happens if there are multiple such symlink replacements in the path? Stuff like that...

Basically what this patch does is it effectively *disables* the check for all self-conflicts within the same package, during test transactions (provided that there also is a %pretrans, but still). So, the worst that can happen if the check detects a valid conflict is that the test transaction passes silently and we only fail before the real one. Given that such self-conflicts are *probably* (I've got no hard data for that claim) pretty rare, it doesn't seem too dangerous. But sure, there is certain risk...

Reply:
Having written the above, I realize that maybe just detecting one case of a symlink replacement in the path would suffice for the conflict exception to be "granted" :)

Reply:
Another way would be to save a flag somewhere after detecting the symlink replacement in the removal check (the first heuristic), and then only trigger the conflict exception in the second heuristic if the file's basename matches the one of the symlink.